### PR TITLE
Fix: Select another pod if all hosts in the pod becomes unavailable

### PR DESCRIPTION
**Pull request by vishesh92**
### Description

In case of a failure while deploying VM, we reset the host_id for the failed VM to null but not the pod_id. This results in failure when there is enough capacity in another pod, but not in the existing pod.

### Types of changes

- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] Enhancement (improves an existing feature and functionality)
- [ ] Cleanup (Code refactoring and cleanup, that may add test cases)
- [ ] build/CI

#### Feature/Enhancement Scale

- [ ] Major
- [ ] Minor

#### Bug Severity

- [ ] BLOCKER
- [ ] Critical
- [ ] Major
- [x] Minor
- [ ] Trivial

#### How did you try to break this feature and the system with this change?
This needs an environment with 2 pods to reproduce the issue and test the fix.

1. On management server, set a debugger here:
https://github.com/apache/cloudstack/blob/9df580cef457cdb767aa5bea926500fa8b1263ca/server/src/main/java/com/cloud/capacity/CapacityManagerImpl.java#L383
2. Deploy a VM. When the debugger reaches the line above, do the following:
    1) Run `SELECT id, state, pod_id, host_id, last_host_id FROM vm_instance ORDER BY id DESC LIMIT 1;` on the `cloud` database.
    2) Get the pod_id from the above and run this query for that pod_id `UPDATE host_pod_ref SET allocation_state = 'Disabled' WHERE id = <pod id>`.
    3) Set `hostHasCpuCapability = false` in the debugger to throw an error in the first run.
    4) VM is retried again once more after this failure. Before the fix, it won't stop at the debugger since it no longer has any available resources to deploy on. After the fix, it will stop again at the debugger. At this point, you can check that pod_id is different.

### CODE REVIEW

File: engine/orchestration/src/main/java/com/cloud/vm/VirtualMachineManagerImpl.java [hunk 1]
```diff
@@ -1460,6 +1466,11 @@ public void orchestrateStart(final String vmUuid, final Map<VirtualMachineProfil
         }
     }
 
+    private boolean checkForNonAllocatedVolumes(long vmId) {
+        final List<VolumeVO> vols = _volsDao.findByInstance(vmId);
+        return CollectionUtils.isEmpty(vols) || vols.stream().allMatch(v -> Volume.State.Allocated.equals(v.getState()));
```
Comment (DaanHoogland) [inline engine/orchestration/src/main/java/com/cloud/vm/VirtualMachineManagerImpl.java:1471]: so false if any are not in  allocated-state? How does this logically is a `checkForNonAllocatedVolumes()`? seems like a `checkThatAllVolumesAreAllocated()` to me.

File: engine/orchestration/src/main/java/com/cloud/vm/VirtualMachineManagerImpl.java [hunk 2]
```diff
@@ -1443,6 +1443,12 @@ public void orchestrateStart(final String vmUuid, final Map<VirtualMachineProfil
                 }
                 if (canRetry) {
                     try {
+                        // Setting pod id to null will result in migration of Volumes across pods
+                        // We set it to null only if migration of volumes across cluster is enabled
+                        // Or volumes are still in allocated state for that VM (in case of failure during deployment)
+                        if (MIGRATE_VM_ACROSS_CLUSTERS.valueIn(vm.getDataCenterId()) || checkForNonAllocatedVolumes(vm.getId())) {
+                            vm.setPodIdToDeployIn(null);
+                        }
```
[the remark this answers is not in the document]
Comment (vishesh92) [reply engine/orchestration/src/main/java/com/cloud/vm/VirtualMachineManagerImpl.java:1451]: Let me update the comment in detail about why we are doing this here.

**Comment by vishesh92** [comment]
@blueorangutan package

**Comment by rohityadavcloud** [comment]
@blueorangutan test matrix

File: engine/orchestration/src/main/java/com/cloud/vm/VirtualMachineManagerImpl.java [hunk 3]
```diff
@@ -1460,6 +1469,11 @@ public void orchestrateStart(final String vmUuid, final Map<VirtualMachineProfil
         }
     }
 
+    private boolean checkThatAllVolumesAreAllocated(long vmId) {
```
Comment (DaanHoogland) [inline engine/orchestration/src/main/java/com/cloud/vm/VirtualMachineManagerImpl.java:1472]: sorry to be a nag (not really) I think 
```suggestion
    private boolean areAllVolumesAllocated(long vmId) {
```
is even better

Comment (DaanHoogland) [reply engine/orchestration/src/main/java/com/cloud/vm/VirtualMachineManagerImpl.java:1472]: or
```suggestion
    private boolean allVolumesAreAllocatedFor(long vmId) {
```
a boolean method starting with `check...()` seems off a bit. I would expect an exception for those `check..`-methods .

**Comment by vishesh92** [comment]
@blueorangutan package

File: engine/orchestration/src/main/java/com/cloud/vm/VirtualMachineManagerImpl.java [hunk 4]
```diff
@@ -1443,6 +1443,15 @@ public void orchestrateStart(final String vmUuid, final Map<VirtualMachineProfil
...
                 if (canRetry) {
                     try {
+                        // Setting pod id to null can result in migration of Volumes across pods. This is not desirable for VMs which
+                        // have a volume in Ready state (happens when a VM is shutdown and started again).
+                        //
+                        // So, we set it to null only when
+                        //   migration of volumes across cluster is enabled
+                        //   Or, volumes are still in allocated state for that VM (happens when VM is Starting/deployed for the first time)
+                        if (MIGRATE_VM_ACROSS_CLUSTERS.valueIn(vm.getDataCenterId()) || areAllVolumesAllocated(vm.getId())) {
+                            vm.setPodIdToDeployIn(null);
```
Comment (DaanHoogland) [inline engine/orchestration/src/main/java/com/cloud/vm/VirtualMachineManagerImpl.java:1453]: shall we make this a method and add the comment as javadoc?

**Review by shwstppr — APPROVED** [review]
Code LGTM

**Comment by DaanHoogland** [comment]
@blueorangutan test

**Comment by DaanHoogland** [comment]
@blueorangutan package

**Comment by vishesh92** [comment]
@blueorangutan package

**Comment by vishesh92** [comment]
@blueorangutan package

**Review by DaanHoogland — APPROVED** [review]
clgtm, I'll be testing it manually to simulate the right conditions.

**Comment by DaanHoogland** [comment]
@blueorangutan package

**Comment by DaanHoogland** [comment]
@blueorangutan test alma9 kvm-alma9

**Comment by DaanHoogland** [comment]
not sure if the errors are related;
@blueorangutan  test alma9 kvm-alma9

**Comment by DaanHoogland** [comment]
@blueorangutan test alma9 kvm-alma9

**Comment by rohityadavcloud** [comment]
JFYI @DaanHoogland @vishesh92 I had some issues using Alma Linux (due to repo/mirror issue) but OL8/OL9 seems to work fine with backend CI/CD.

**Comment by DaanHoogland** [comment]
@blueorangutan  test alma9 kvm-alma9

**Comment by DaanHoogland** [comment]
> @blueorangutan test alma9 kvm-alma9

this didn´t work :exploding_head: , so started it manually

**Comment by DaanHoogland** [comment]
>> @blueorangutan test alma9 kvm-alma9

> this didn´t work :exploding_head: , so started it manually

results:
Smoke tests completed. 108 look OK, 1 have errors, 0 did not run
Only failed and skipped tests results shown below:


Test | Result | Time (s) | Test File
--- | --- | --- | ---
test_01_migrate_VM_and_root_volume | `Error` | 89.26 | test_vm_life_cycle.py
test_02_migrate_VM_with_two_data_disks | `Error` | 51.66 | test_vm_life_cycle.py
test_08_migrate_vm | `Error` | 0.07 | test_vm_life_cycle.py

These error are all over the place at the moment, not specific to this issue.

**Comment by DaanHoogland** [comment]
tested according to spec in the description.

**Pull request merged**